rfcs/0046-versioning-changes-for-hard-fork.md

```diff
@@ -0,0 +1,169 @@
+## Summary
+[summary]: #summary
+
+Remove unneeded version tags from `Bin_prot`-serialized data produced
+by `ppx_version`. Generate version tags on an as-needed basis where
+they're useful. Detect versioning violations in the CI version linter
+in a more useful way.
+
+## Motivation
+[motivation]: #motivation
+
+Versioned data types are built from other versioned types and
+primitive OCaml types. When generating `Bin_prot`-serialized data from
+instances of those type, a version tag is generated not only for the
+containing type, but also for each constituent type. That strategy
+allows determining the version of any piece of the serialized data.
+
+The Jane Street `Async` versioned RPC mechanism used in
+`Mina_networking` already provides versioning of data sent between
+nodes. Therefore, version-tagging of the serialized data is not needed
+for that use case. Removing the version tags will reduce the amount of
+data sent over the network.
+
+We may wish to save serialized data and be able to determine its
+version. For that use case, a single "top" tag for the containing type
+suffices.
+
+Some user-visible data, such as public keys, are formatted using the
+current serialization mechanism, where all contained types have
+version tags. For those few cases, allow use of the existing
+mechanism.
+
+The CI version linter currently compares changes to versioned types
+against the PR base branch. That's more fine-grained than needed, if
+the goal of preventing version changes is to allow interoperability
+with deployed software. Instead, compare the types against the branch
+of the latest released software. That strategy will reduce the amount
+of noise produced by the linter, and provide more useful results.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Version-tagging changes
+
+Currently, tagging a `Stable` module with `%versioned` produces
+shadowing `bin_read_t` and `bin_write_t` functions that read and write
+version tags, positive integers, prepended to serialized data, via the
+`ppx_version` library. The proposal is not to shadow those functions,
+and instead use those generated by `deriving bin_io`, directly.  It is
+still useful to use `%versioned` to assure that serializations do not
```

Comment:
Would be nice to get rid of `versioned` entirely. I hope https://github.com/MinaProtocol/mina/issues/11101 may be able to help with this.

Reply:
If you can work out how to do that, while maintaining the stability of serialized types, that'd be great.

Reply:
I was hoping that #11101 would eventually give us a place to check that types don't change between versions.